### ISSUE TITLE
Grails 3.0: When app has context path, url mappings only working for '/context/' with trailing slash when spring security implemented.

### ISSUE THREAD

**Issue by kschmit90** [grails/grails-core#9024]
Simply, I have a standard Spring Security authentication configuration, which works fine when I am not using a context path for the server. 

````
environments:
    development:
        server: 
            contextPath: '/context'

````

When I add a context path, my assumption is that a URL like 'localhost:8080/context' should map to whatever is mapped to "/" in UrlMappings. In this instance it is the index action of a controller.

````
class UrlMappings {

  static mappings = {
    "/$controller/$action?/$id?(.$format)?"{
      constraints {
        // apply constraints here
      }
    }

    "/"(controller:'myController', action:'index')
    "500"(view:'/error')
    "404"(view:'/notFound')
  }
}
````

This is true when not using Spring Security to place a log in form in front of whatever URL is being requested. 

The security configuration looks like this. It is very basic, block some resources, permit viewing assets, then make sure everything else is authenticated and place a log in form in front of not authenticated requests.

````
class SecurityConfiguration extends WebSecurityConfigurerAdapter {

  @Override
  protected void configure(HttpSecurity http) throws Exception {
    http
      .authorizeRequests()
        .antMatchers("/assets/**").permitAll()
        .antMatchers("/category/**").hasRole("ADMIN")
        .antMatchers("/subCategory/**").hasRole("ADMIN")
        .antMatchers("/audience/**").hasRole("ADMIN")
        .antMatchers("/location/**").hasRole("ADMIN")
        .anyRequest()
        .authenticated()
        .and()
      .formLogin()
        .loginPage('/authentication/login')
        .permitAll()
        .and()
      .logout()
        .logoutUrl("/authentication/logoutRedirect")
        .invalidateHttpSession(true)
        .permitAll()
  }

````

The issue occurs when there is a context path. If a URL like 'localhost:8080/context' is used, and authentication is successful, the application does NOT proceed to "/"(controller:'knowledgeAsset', action:'index') as expected. 

However, when a URL like 'localhost:8080/context/' with the explicitly added trailing slash is used, and authentication is successful, the application DOES proceed to "/"(controller:'knowledgeAsset', action:'index') as expected.

**Comment by graemerocher** [comment]
Could you upload an example that reproduces the problem?

**Comment by kschmit90** [comment]
Example application reproducing the issue.

https://github.com/kschmit90/grails3contextPathSecurityTest

**Comment by graemerocher** [comment]
@rwinch Any idea why login doesn't work with Spring Security and Grails when there is no trailing slash / in the URI? It doesn't seem to even hit the Grails layer, so seems a Spring Security config issue...

**Comment by rwinch** [comment]
# The Issue

## Summary

The issue can be summarized by the fact that the `HttpServletRequest.getRequestURI()` is returning `/context` instead of `/context/`. This cause the session information to be lost due to a "quirk" in Tomcat and the configuration of `GrailsDispatcherServlet`.

## Details

The details are as follows:

* When a user requests `/context` Spring Security's `ExceptionTranslationFilter` uses the `RequestCache` to save the current request. The request that is returned from Tomcat is one that has a requestURI of `/context`. This can be observed by adding a breakpoint inside the constructor of `DefaultSavedRequest`
* After a user authenticates successfully, Spring Security's `SimpleUrlAuthenticationSuccessHandler` is used to determine the originally requested URL. This is done by consulting the `DefaultSavedRequest` which uses the requestURI to construct a redirect of http://localhost:8080/context
* A redirect is sent to `http://localhost:8080/context` for the request URL of http://localhost:8080/context/login which includes a JSESSIONID cookie that has a path of `/context/`. The authenticated user is associated to that session.
* The browser requests http://localhost:8080/context which will not include the JSESSIONID cookie since the path of the cookie is `/context/` (NOTE the trailing slash) and the request path is `/context` (NOTE no trailing slash).
* Spring Security will not see the session because the JSESSIONID cookie was not included. This means the user is not authenticated and thus not authorized to view `/context`.
* Spring Security again saves the request of `/context` to the (new) session. The session is new because no JSESSIONID cookie was submitted with the request.
* Spring Security sends a redirect to the login page with a new JSESSIONID cookie since a new session was created to persist the saved request.

## Jetty

This issue will not exist in Jetty because Jetty does not have the Filter intercept the URL `/context`.  Also Jetty maps the JSESSIONID cookie to `/context` instead of `/context/`

## Simplified Example

You can find a simplified example of the problem at https://github.com/rwinch/tomcat-context-trailing-slash

# Fixing the Issue

## Remapping GrailsDispatcherServlet

Tomcat will allow intercepting the URL `/context` when a servlet is mapped to `/*`. To fix it, the `GrailsDispatcherServlet` can be mapped to `/` (default servlet) instead. By mapping to the default servlet, Tomcat itself will perform a redirect from `/context` to `/context/` so the application code will not be executed and avoid this bug.

##  sessionCookiePathUsesTrailingSlash

Alternatively (but not recommended), you can configure Tomcat to set `sessionCookiePathUsesTrailingSlash` to `false`. See the [Tomcat Docs|https://tomcat.apache.org/tomcat-8.0-doc/config/context.html#Common_Attributes] for how to do this and why it is not a good idea.

# TBD

I'm currently in talks with the Tomcat team to see if the the difference in behavior for mapping the servlet to `/` vs `/*` is expected and if there are changes that should be made to Tomcat.

**Comment by graemerocher** [comment]
@rwinch Wow. Thanks for the fantastic thorough answer and response :)

**Comment by kschmit90** [comment]
The problem still exists in 3.0.3, I have tested with the same sample project submitted above, in a grails 3.0.3 environment.

**Comment by graemerocher** [comment]
Yes didn't have time to fit it into 3.0.3 (due to lack of time to properly test it), but it is fixed in 3.0.4.BUILD-SNAPSHOT

**Comment by rwinch** [comment]
@kschmit90 Another way to workaround it would be to add a Filter before Spring Security and if the path is equal to "/context" to redirect to "/context/". This ensures that Spring Security will not try to save the request to "/context". This might help until 3.0.4 is released.

**Comment by tcrespog** [comment]
Hi,

This issue reproduces again with Grails 3.1.8. In 3.1.7 it just works fine. It can be verified with the original sample application: https://github.com/kschmit90/grails3contextPathSecurityTest

I believe [this commit](https://github.com/grails/grails-core/commit/2c45566264761660e9521b498774be111e907a53) introduced this issue again. Could it be the case?

**Comment by graemerocher** [comment]
It seems JBoss doesn't like `/*` and Tomcat doesn't like `/`. Oh the joys of servlet containers. Try setting `grails.web.servlet.path` to `/` in `application.yml`

**Comment by graemerocher** [comment]
@rwinch Anything of interest emerge from your talks with the Tomcat team? It seems right now each container needs tweaks to the servlet path mappings to make this work

**Comment by rwinch** [comment]
@graemerocher Sorry for not following up sooner. I spoke with @markt-asf offline. I have provided some quotes from our conversation (with his permission below).

# tl;dr

The short answer is that there is nothing Tomcat can do because the bug that Tomcat is working around is still present in IE11. As for a workaround he recommends:

> I'd go with the work-around of setting sessionCookiePathUsesTrailingSlash="false" for 99.9% of cases. It is really only a problem if you don't control / trust the other web applications deployed on the container instance. The threat here is a malicious web app could steal the session ID allowing an attacker to impersonate the user. If you control / trust all the apps, that is a non-issue.

# Details

Here are relevant parts of our conversation. The questions are from me and the responses are from @markt-asf with his permission.

## Why the behavior is happening?

> The behaviour you are describing is documented in the docs for sessionCookiePathUsesTrailingSlash although I need to check if other browsers send the cookie for /foo/ if the request is to /foo (if they don't the text will need some tweaking).
>
> This feature is a workaround for what it arguably a bug in IE. I haven't checked newer versions of IE to see if this is still a problem or not. If IE has addressed this then we can simply remove this feature.
>
> If IE still has this issue I'm not sure what we can do. There is no way to protect against the IE bug and have the session available for a request to /foo if a servlet is mapped to /*. You effectively have to pick one or the other. i.e. live with the potential IE bug (likely to be acceptable if you control what other contexts will be installed on that server) or not have a servlet mapped to /*.

...

> IE11 is still broken (my previous test was bad) so there is no scope for removing the workaround.

## Why does the example work when the servlet is mapped to "/" (default servlet) vs "/*"?

> Because of optimisations in the code that maps requests to hosts, contexts and servlets.
>
> In the / case, Tomcat knows that /context will not map to a servlet and will therefore be handled by the default servlet which will issue a 302 to /context/. Tomcat therefore issues the redirect from the mapper to get a faster response to the client. This means those requests bypass any filters.

## Why is there a difference between "/*" and "/"

If the servlet is mapped to `"/*"` and a Filter is mapped to `"/*"`, then the Filter is invoked for `"/context"`. To me this discrepancy is quite subtle (and seems like the bug). If this were changed to work the same as when the servlet is mapped to "/" (i.e. Tomcat handles the redirect external of user code), then I think it would resolve the issue.

> On the redirect and filtering point the Servlet spec document is far from clear. I've been advocating (withn only limited sucess) to get some clarity added to the spec in various places. It is far from helpful for end-users but the 'definition' of the servlet specification is:
> - if it is clear in the spec document then use that
> - if not, look to the Javadoc for clarification
> - if still unclear what does the TCK do (teh TCK not being available to most people really doesn't help here and due to the ASF's issues with the JCP I don't have access to the latest versions)
> - if still unclear is there anything in the EG mailing lists that describes the intent of the EG
> 
> On this particular point the information is spread far and wide.
> 
> The first point is that the redirect from /foo to /foo/ is 'defined' in section 10.10 (welcome files) where in the examples towards the end of that section it shows a couple of cases where this redirect takes place. The implication is that this redirect is part of welcome file processing although no clear requirement is stated anywhere else I can find that this redirect should take place.
> 
> There was a discussion in the servlet 3.1 EG mailing lists about whether welcome fie processing took place before or after filter mapping. There was a clear consensus in the EG that it was before although I'm not sure if that ever turned into actual clarification text in the spec.
> 
> Put these together and you get Tomcat's current behaviour.
> 
> A servlet mapping of '/*' means no welcome file processing is necessary because the request will always match the servlet mapped to '/*'. And that is why you get the difference in behaviour between '/' (welcome file processing including the redirect takes place before the FilterChain) and '/*' (no welcome file processing so it reaches the FilterChain).
> 
> In summary, it is all a bit of a mess but as far as I can tell Tomcat is implementing what the spec requires.
> 
> It is a tad frustrating that the problem you are seeing is caused by a workaround to an IE issue in that IE doesn't follow the requirements of RFC6265 , section 4.1.2.4. I'd raise a bug with Microsoft but, having spent a couple of weeks trying to get them to accept a simple WebDAV client bug, I very much doubt I'd get anywhere with this IE bug.

Hope this helps!

**Comment by tcrespog** [comment]
@graemerocher @rwinch Thank you both. The grails.web.servlet.path workaround also seems to work.

**Comment by graemerocher** [comment]
@rwinch  @markt-asf This seems like a massive bug to me. I hope a solution is found in Tomcat soon because right now it is not possible to create a WAR that deploys to both JBoss and Tomcat without adding custom logic that changes the mapping from `/` to `/*` in JBoss. I imagine this impacts other containers as well.

**Comment by markt-asf** [comment]
The main problem here is the original IE cookie path handling bug, Microsoft's historic reluctance to accept and fix bug reports and Tomcat's decision, therefore, to work-around it by default. Given the IE bug has - in rare cases - security implications Tomcat's decision seems reasonable but it is causing problems and those problems look to be more frequent than the security issues so it is, in my view, worth revisiting.

Regarding the differences in behavior between Tomcat and JBoss, a large factor in that is the lack of clarity in the Servlet spec. I suspect if you asked either the Tomcat or the JBoss team both would assert the behavior of their respective container's was spec compliant and both would likely be right. As I pointed out, I am trying to get the spec clarified but despite being on the Servlet EG, all I can do is lobby Oracle to make changes. Even if every other member of the EG agrees with me, there is no way to make the change happen unless Oracle wants to do it.

Having Tomcat do the `/context` to `/context/` redirect when there is a Servlet mapped to `/*` would be a breach of the specification since the Servlet mapping matches the request for `/context` so the Servlet should handle it. In theory, an application could respond differently for `/context` and `/context/` although in practice such differences are rare. Given this, I don't think having Spring Security always do this redirect is a good idea as it would break those rare cases.

You can configure Tomcat not to do the redirect from `/context` to `/context/` early (when no servlets are mapped to `/*`) via the `mapperContextRootRedirectEnabled` attribute of the `Context` but that would actually make the problem worse, not better, in this case.

I need to re-check browser behavior for the cookie handling and then I'll start a discussion on the Tomcat dev list.

**Comment by markt-asf** [comment]
The conclusion on the Tomcat dev list was that the problems caused by enabling `sessionCookiePathUsesTrailingSlash` by default are greater than the security risks associated with disabling by default so the default will be changed for the next set of releases.

**Comment by graemerocher** [comment]
I've pushed a fix for this to Grails until a version of Tomcat exists that resolves this fully

**Comment by rodrigosalinas** [comment]
I'm having the same issue now with Grails 3.3.6. What I understand is that I have to set the sessionCookiePathUsesTrailingSlash to true? Is that correct? Thanks in advance.

**Comment by chongtin** [comment]
This problem happen with Grails 3.3.6 on undertow.version '1.4.18.Final' as well.
Like  graemerocher said, setting **grails.web.servlet.path to / in application.yml** does solve the problem.

**Comment by stlserverless** [comment]
FYI: We've had the servlet.path = / and got the expected results outlined above.  However, after we upgraded an app using Spring Security to Grails 3.3.8, we saw different behavior as the request URI when navigating to localhost:8080/foo resulted in the DefaultUrlMappingsHolder trying to find a mapping for uri = '' (empty string). Since no filterChains mapped that URI, I kept getting redirected to a login endpoint. I tried changing servlet.path back to /*, but got the same result.

However, after changing `contextPath: '/foo'` to `contextPath: '/foo/'` with a trailing slash, we're again seeing the default view for / rendered instead of getting the redirect to the login.  Am still investigating to see if there are any undesirable side-effects, but so far, so good.

**Comment by jfrosch** [comment]
FYI: We've had the servlet.path = / and got the expected results outlined above. However, after we upgraded an app using Spring Security to Grails 3.3.8, we saw different behavior as the request URI when navigating to localhost:8080/foo resulted in the DefaultUrlMappingsHolder trying to find a mapping for uri = '' (empty string). Since no filterChains mapped that URI, I kept getting redirected to a login endpoint. I tried changing servlet.path back to /*, but got the same result.

However, after changing contextPath: '/foo' to contextPath: '/foo/' with a trailing slash, we're again seeing the default view for / rendered instead of getting the redirect to the login. Am still investigating to see if there are any undesirable side-effects, but so far, so good.

**Comment by 1HzAkLAT** [comment]
i am seeing an issue inside WebLogic server, so added grails.web.servlet.path="/" inside external config file, but its not solving the problem.

ISSUE: in weblogic server http://host:8080/myapp is not getting redirected to **http://host:8080/myapp/** (not adding trailing slash at the end)

**Comment by rmarquezromero** [comment]
> i am seeing an issue inside Grails 3.3.2 - WebLogic server, so added grails.web.servlet.path="/" inside external config file, but its not solving the problem.
> 
> ISSUE: in weblogic server http://host:8080/myapp is not getting redirected to **http://host:8080/myapp/** (not adding trailing slash at the end)

Hi @vhalthi 

Finally did you find a solution? I have same problem.

Grails 3.3.11
Weblogic 14.1.1

**Comment by rmarquezromero** [comment]
> I've pushed a fix for this to Grails until a version of Tomcat exists that resolves this fully

Hi Graeme!

Is this solved for Weblogic 14?

**Comment by rmarquezromero** [comment]
> I've pushed a fix for this to Grails until a version of Tomcat exists that resolves this fully



> I've pushed a fix for this to Grails until a version of Tomcat exists that resolves this fully

Hi @graemerocher !

Is this solved for Weblogic 14?
